framework/src/arastorage/relation.c

```diff
@@ -991,8 +1015,24 @@ db_result_t relation_process_remove(db_handle_t **handle, db_cursor_t *cursor)
 		memcpy(result_row + attr_map_ptr->to_offset, from_ptr, from_attr->element_size);
 	}
 
+	not_remove = ((*handle)->lvm_instance == NULL || lvm_execute((*handle)->lvm_instance) == FALSE);
+	{
+		from_attr = list_head((*handle)->rel->attributes);
+		while (from_attr != NULL) {
+			if (from_attr->index != NULL) {
+				if (relation_get_value((*handle)->rel, from_attr, row, &index_key) == DB_OK) {
+					index_delete(from_attr->index, &index_key);
+					if (not_remove) { //update with new tuple_id
+						index_insert(from_attr->index, &index_key, (*handle)->result_rel->cardinality);
+					}
+				}
+			}
+			from_attr = from_attr->next;
+		}
+	}
+
 	/* Check whether the given predicate is true for this tuple. */
-	if ((*handle)->lvm_instance == NULL || lvm_execute((*handle)->lvm_instance) == FALSE) {
+	if (not_remove) {
 		result = storage_put_row((*handle)->result_rel, result_row, TRUE);
 		if (DB_ERROR(result)) {
```

Comment:
If a index is added in line 1026 and then this error happens, index information is wrong,
I think you need to handle this case.

Reply:
refined the code logic